base_optional_quick_create/model.py

```diff
@@ -50,14 +48,18 @@ def _register_hook(self, cr, ids=None):
                     model_obj.check_quick_create = True
         return True
 
+    def _register_hook(self, cr):
+        self._patch_quick_create(cr, self.search(cr, SUPERUSER_ID, []))
+        return super(IrModel, self)._register_hook(cr)
+
     def create(self, cr, uid, vals, context=None):
-        res_id = super(ir_model, self).create(cr, uid, vals, context=context)
-        self._register_hook(cr, [res_id])
+        res_id = super(IrModel, self).create(cr, uid, vals, context=context)
+        self._patch_quick_create(cr, [res_id])
```

Comment:
Still it does not seem a good trigger point to do it. If you quick-create a record before slow-creating one, it would still enable it.

The logical flow IMHO should be like:

```
_inherit = 'ir.model'
@api.cr
def _register_hook(self):
  self.search([("avoid_quick_create", "=", True)])._patch_quick_create()
  return super(...)._register_hook()

@api.multi
def _patch_quick_create(self):
  @api.model
  def new_name_create_method(self):
    ...
  for s in self:
    self.env[s.model]._patch_method("name_create", new_name_create)
```

And add a constraint to `avoid_quick_create` column that patches or restores the method depending on what changed.